Add Cosmwasm Optimint Tutorial (#151)
celestiaorg/docs: docs/developers/cosmwasm-contract-deployment.md

# Contract Deployment on CosmWasm with Optimint
<!-- markdownlint-disable MD013 -->

## Compile the Smart Contract

We will run the following commands to pull down the Nameservice
smart contract and compile it:

```sh
git clone https://github.com/InterWasm/cw-contracts
cd cw-contracts

> git checkout main
> cd contracts/nameservice
> rustup default stable

Reviewer: `main` is the default branch and `rustup` was configured when setting up dependencies.

```suggestion
cd contracts/nameservice
```

> RUSTFLAGS='-C link-arg=-s' cargo wasm

Reviewer: Are the flags actually needed? Built fine for me without them.

```suggestion
cargo wasm
```

```

> The compiled contract is outputted to the following repository:

Reviewer: ```suggestion
The compiled contract is outputted to:
```

`target/wasm32-unknown-unknown/release/cw_nameservice.wasm`.

## Unit Tests

If we want to run tests, we can do so with the following command:

```sh

> RUST_BACKTRACE=1 cargo unit-test

Reviewer: ```suggestion
cargo unit-test
```

```

## Optimized Smart Contract

Because we are deploying the compiled smart contract to `wasmd`,
we want it to be as small as possible.

CosmWasm team provides a tool called `rust-optimizer` which we need
Docker for in order to compile.

Run the following command:

```sh
docker run --rm -v "$(pwd)":/code \
  --mount type=volume,source="$(basename "$(pwd)")_cache",target=/code/target \
  --mount type=volume,source=registry_cache,target=/usr/local/cargo/registry \
  cosmwasm/rust-optimizer:0.12.6
```

> This will compile our code inside `artifacts/cw_nameservice.wasm` directory.

Reviewer: ```suggestion
This will place the optimized Wasm bytecode at `artifacts/cw_nameservice.wasm`.
```
